### PR TITLE
Bigger fonts and highlight starts on 2nd line

### PR DESCRIPTION
**Pull request by pedrosino**
Improved fitText to allow bigger fonts. Also made the highlighting start on the 2nd line instead of the 1st (less lines highlighted)

### CODE REVIEW

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 1]
```diff
@@ -92,6 +92,7 @@
     private static final BaseColor HIGHLIGHT_COLOR = new BaseColor(230, 230, 230);
     private static final int SCRAMBLE_PADDING_VERTICAL = 3;
     private static final int SCRAMBLE_PADDING_HORIZONTAL = 3;
+    private static final int SCRAMBLE_LEADING = 16;
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Is this used anywhere?

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 2]
```diff
@@ -803,11 +849,13 @@ private static TableAndHighlighting createTable(PdfWriter docWriter, Document do
             // Again, I have no idea where this number is coming from. I'm chalking it up to
             // unaccounted for margins.
             int WIDTH_MARGINS = 40;
-            Rectangle availableArea = new Rectangle(scrambleColumnWidth - WIDTH_MARGINS,
+            Rectangle availableArea = new Rectangle(scrambleColumnWidth - 2*SCRAMBLE_PADDING_HORIZONTAL/* - WIDTH_MARGINS*/,
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: What's going on here? Why the commented out code?

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 3]
```diff
@@ -803,11 +849,13 @@ private static TableAndHighlighting createTable(PdfWriter docWriter, Document do
...
                     availableScrambleHeight - HEIGHT_MARGINS);
+            System.out.println("Normal..." + scrambler.getShortName());
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Please no print statements in the final version of tnoodle. If you really need something, there's a logging system that lets us turn these things on and off as needed without recompiling tnoodle.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 4]
```diff
@@ -803,11 +849,13 @@ private static TableAndHighlighting createTable(PdfWriter docWriter, Document do
...
             float perfectFontSize = fitText(new Font(courier), longestPaddedScramble, availableArea, MAX_SCRAMBLE_FONT_SIZE, PdfWriter.RUN_DIRECTION_LTR, true);
             if(tryToFitOnOneLine) {
                 String longestScrambleOneLine = longestScramble.replaceAll(".", widestCharacter + "");
+                System.out.println("OneLine..." + scrambler.getShortName());
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: print statement

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 5]
```diff
@@ -815,6 +863,8 @@ private static TableAndHighlighting createTable(PdfWriter docWriter, Document do
                 }
             }
             scrambleFont = new Font(courier, perfectFontSize, Font.NORMAL);
+            float scHeight = scrambleFont.getBaseFont().getAscentPoint("R", perfectFontSize) - scrambleFont.getBaseFont().getDescentPoint("R", perfectFontSize);
+            System.out.println("Font: " + perfectFontSize + " = height " + scHeight);
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: print statement

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 6]
```diff
@@ -815,6 +863,8 @@ private static TableAndHighlighting createTable(PdfWriter docWriter, Document do
                 }
             }
             scrambleFont = new Font(courier, perfectFontSize, Font.NORMAL);
+            float scHeight = scrambleFont.getBaseFont().getAscentPoint("R", perfectFontSize) - scrambleFont.getBaseFont().getDescentPoint("R", perfectFontSize);
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: What's this for?

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 7]
```diff
@@ -831,11 +881,12 @@ private static TableAndHighlighting createTable(PdfWriter docWriter, Document do
             table.addCell(nthscramble);
 
             Phrase scramblePhrase = new Phrase();
-            boolean oddLine = false;
+            boolean oddLine = true;
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Why did you change this? Line 0 is an even line.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 8]
```diff
@@ -620,60 +621,104 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
-            ct = new ColumnText(null);
-            ct.setSimpleColumn(ph, rect.getLeft(), rect.getBottom(), rect.getRight(), rect.getTop(), maxFontSize, Element.ALIGN_LEFT);
-            ct.setRunDirection(runDirection);
-            status = ct.go(true);
-            if ((status & ColumnText.NO_MORE_TEXT) != 0 && (newlinesAllowed || ct.getLinesWritten() <= 1)) {
-                return maxFontSize;
-            }
+
+        //we don't want the text to go all the way to the top
+        float maxHeight = rect.getHeight() - 8.0f;
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: I see a lot of hardcoded numbers in this function like this `0.8f`. I'm not going to try reading this code until they're all gone =).

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 9]
```diff
@@ -799,11 +840,12 @@ private static TableAndHighlighting createTable(PdfWriter docWriter, Document do
             // Gah, I have no idea where this number is coming from, see
             // https://github.com/cubing/tnoodle/issues/124 for why we had to bump
             // it from 8 to 9.
-            int HEIGHT_MARGINS = 9;
+            // Pedro says: I changed this margin to 17, so that text doesn't go all the way to the top of the cell and touch the border
+            int HEIGHT_MARGINS = 17;
             // Again, I have no idea where this number is coming from. I'm chalking it up to
             // unaccounted for margins.
             int WIDTH_MARGINS = 40;
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: What is `WIDTH_MARGINS` used for?

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 10]
```diff
@@ -799,11 +840,12 @@ private static TableAndHighlighting createTable(PdfWriter docWriter, Document do
             // Gah, I have no idea where this number is coming from, see
             // https://github.com/cubing/tnoodle/issues/124 for why we had to bump
             // it from 8 to 9.
-            int HEIGHT_MARGINS = 9;
+            // Pedro says: I changed this margin to 17, so that text doesn't go all the way to the top of the cell and touch the border
+            int HEIGHT_MARGINS = 17;
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: The original comment seems to be vestigial now, lets remove it. I think your comment doesn't add much either (you're just defining what a margin is, so lets remove it as well).

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 11]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
-                return maxFontSize;
-            }
+
+        //first verify if there's a line break ("\n") -> megaminx
+        //if so, calculate the font based only on the width
+        int lineBreak = text.indexOf("\n");
+
+        if(lineBreak > 0) {
+            //the char width is fontSize * 0.6 -> this was found with getWidthPoint()
+            float size = rect.getWidth() / (lineBreak * 0.6f);
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Lets use `getWidthPoint()` instead of commenting that that's how you found it.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 12]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
+
+            //the calculated size is a little too big, so we reduce it to 95%
+            return (size * 0.95f);
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Rather than multiplying by 0.95 here, lets increase the margins in the code that calls `fitText`.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 13]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
-            ct = new ColumnText(null);
-            ct.setSimpleColumn(ph, rect.getLeft(), rect.getBottom(), rect.getRight(), rect.getTop(), maxFontSize, Element.ALIGN_LEFT);
-            ct.setRunDirection(runDirection);
-            status = ct.go(true);
-            if ((status & ColumnText.NO_MORE_TEXT) != 0 && (newlinesAllowed || ct.getLinesWritten() <= 1)) {
-                return maxFontSize;
-            }
+
+        //first verify if there's a line break ("\n") -> megaminx
+        //if so, calculate the font based only on the width
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Don't we need to look at the height as well?

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 14]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
+            //the char width is fontSize * 0.6 -> this was found with getWidthPoint()
+            float size = rect.getWidth() / (lineBreak * 0.6f);
+
+            //the calculated size is a little too big, so we reduce it to 95%
+            return (size * 0.95f);
+        }
+
+        //if we can use more than one line,
+        if(newlinesAllowed) {
+            float minFont = 8.0f;
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Hmmm... lets pull this out into a constant at the top of the file.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 15]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
-                ct.setSimpleColumn(new Phrase(text, font), rect.getLeft(), rect.getBottom(), rect.getRight(), rect.getTop(), size, Element.ALIGN_LEFT);
-                ct.setRunDirection(runDirection);
-                status = ct.go(true);
-                if ((status & ColumnText.NO_MORE_TEXT) != 0 && (newlinesAllowed || ct.getLinesWritten() <= 1)) {
-                    if (max - min < size * precision) {
-                        return size;
-                    }
-                    min = size;
+
+            //first make the chunks using a potentialFontSize
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: What's the `potentialFontSize`  that this comment is referring to?

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 16]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
+            //starting with the maximum size
+            float oldFont = minFont;
+            float newFont = maxFont;
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Can we rename `newFont` to `potentialFontSize`, and have it start in the middle of `minFontSize` and `maxFontSize`?

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 17]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
+
+            //determine line height - using font ascent and descent
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Could you add a link to where you got the arithmetic from? It's still not clear to me why it would work.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 18]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
+            float lineHeight = font.getBaseFont().getAscentPoint(text, newFont) - font.getBaseFont().getDescentPoint(text, newFont);
+            
+            font.setSize(newFont);
+
+            LinkedList<Chunk> lineChunks = splitScrambleToLineChunks(text, font, rect.getWidth());
+
+            int nLines = lineChunks.size();
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: `nLines` -> `lineCount` please

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 19]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
+
+                newFont = (float)(maxFont + minFont) / 2.0f;
+
+                lineHeight = font.getBaseFont().getAscentPoint(text, newFont) - font.getBaseFont().getDescentPoint(text, newFont);
+
+                font.setSize(newFont);
+
+                lineChunks = splitScrambleToLineChunks(text, font, rect.getWidth());
+
+                nLines = lineChunks.size();
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: `nLines` -> `lineCount` please.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 20]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
+            float diff = Math.abs(newFont - oldFont);
+
+            boolean stop = false;
+
+            if(diff < precision && fits) {
+                stop = true;
+            }
+
+            //loop until it's good enough
+            while(!stop) {
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: I see a lot of duplicated code here. Lets clean this up. This feels like a good candidate for a do...while loop.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 21]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
+                nLines = lineChunks.size();
+
+                totalHeight = (float)nLines * lineHeight * 1.5f - (0.5f * lineHeight);
+
+                fits = (totalHeight < rect.getHeight());
+
+                diff = Math.abs(newFont - oldFont);
+
+                if(diff < precision && fits) {
+                    stop = true;
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Do we have to worry about infinite loops here if someone passes in a string that simply cannot fit? Perhaps we should cap the iteration at 100 or something?

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java [hunk 22]
```diff
@@ -620,60 +620,100 @@ private static void addScrambles(PdfWriter docWriter, Document doc, ScrambleRequ
...
-            return size;
-        } catch (Exception e) {
-            throw new com.itextpdf.text.ExceptionConverter(e);
+
+            return newFont;
+           
+        } else { //trying to fit in a single line
+            //Again, char width is fontSize * 0.6, but it may get too big (skewb specially)
+            //so we use 0.65 instead
+            float fontSize = (rect.getWidth() / (float)text.length()) * 0.65f;
```
Comment (jfly) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: What's going on here? The comment doesn't make it clear to me why you need 0.65 here instead of 0.6.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java (hunk 14 above)
Comment (pedrosino) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Done. Precision is up there, too.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java (hunk 15 above)
Comment (pedrosino) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: It was an old variable I was using.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java (hunk 16 above)
Comment (pedrosino) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: I prefer oldFont and newFont. Also, we _first_ check if the maxFontSize is small enought, and _then_ start the binary search.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java (hunk 18 above)
Comment (pedrosino) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Done.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java (hunk 20 above)
Comment (pedrosino) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: See a couple comments above. First check maxFont, then start binary search.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java (hunk 19 above)
Comment (pedrosino) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Done.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java (hunk 22 above)
Comment (pedrosino) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Clarified (hopefully)

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java (hunk 10 above)
Comment (pedrosino) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Done.

File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java (hunk 9 above)
Comment (pedrosino) [inline webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/ScrambleRequest.java]: Nothing. Removed
